cpu/sam0_common/periph/i2c.c

```diff
@@ -48,148 +50,110 @@
-#endif
+    sercom_clk_en(dev(bus));
 
     /* I2C using CLK GEN 0 */
+    sercom_set_gen(dev(bus),i2c_config[bus].gclk_src);
 #if CPU_FAM_SAML21
-    GCLK->PCHCTRL[sercom_gclk_id].reg = (GCLK_PCHCTRL_CHEN |
-                         GCLK_PCHCTRL_GEN_GCLK0  );
-    while (GCLK->SYNCBUSY.bit.GENCTRL) {}
-
-    GCLK->PCHCTRL[sercom_gclk_id_slow].reg = (GCLK_PCHCTRL_CHEN |
-                         GCLK_PCHCTRL_GEN_GCLK0  );
+    /* GCLK_ID_SLOW is shared for SERCOM[0..4] */
+    GCLK->PCHCTRL[(sercom_id(dev(bus)) < 5 ?
+                  SERCOM0_GCLK_ID_SLOW : SERCOM5_GCLK_ID_SLOW)].reg =
+    (GCLK_PCHCTRL_CHEN | i2c_config[bus].gclk_src  );
     while (GCLK->SYNCBUSY.bit.GENCTRL) {}
 #else
+    /* GCLK_SERCOMx_SLOW is shared for all sercom */
     GCLK->CLKCTRL.reg = (GCLK_CLKCTRL_CLKEN |
-                         GCLK_CLKCTRL_GEN_GCLK0 |
-                         GCLK_CLKCTRL_ID(sercom_gclk_id));
-    while (GCLK->STATUS.bit.SYNCBUSY) {}
-
-    GCLK->CLKCTRL.reg = (GCLK_CLKCTRL_CLKEN |
-                         GCLK_CLKCTRL_GEN_GCLK0 |
-                         GCLK_CLKCTRL_ID(sercom_gclk_id_slow));
+                         i2c_config[bus].gclk_src |
+                         SERCOM0_GCLK_ID_SLOW);
     while (GCLK->STATUS.bit.SYNCBUSY) {}
 #endif
 
-
     /* Check if module is enabled. */
-    if (I2CSercom->CTRLA.reg & SERCOM_I2CM_CTRLA_ENABLE) {
+    if (dev(bus)->CTRLA.reg & SERCOM_I2CM_CTRLA_ENABLE) {
         DEBUG("STATUS_ERR_DENIED\n");
         return -3;
     }
     /* Check if reset is in progress. */
-    if (I2CSercom->CTRLA.reg & SERCOM_I2CM_CTRLA_SWRST) {
+    if (dev(bus)->CTRLA.reg & SERCOM_I2CM_CTRLA_SWRST) {
         DEBUG("STATUS_BUSY\n");
         return -3;
     }
 
     /************ SERCOM PAD0 - SDA and SERCOM PAD1 - SCL *************/
-    gpio_init_mux(pin_sda, mux);
-    gpio_init_mux(pin_scl, mux);
+    gpio_init_mux(i2c_config[bus].sda_pin, i2c_config[bus].mux);
+    gpio_init_mux(i2c_config[bus].scl_pin, i2c_config[bus].mux);
 
     /* I2C CONFIGURATION */
-    while (I2CSercom->SYNCBUSY.reg & SERCOM_I2CM_SYNCBUSY_MASK) {}
+    while (dev(bus)->SYNCBUSY.reg & SERCOM_I2CM_SYNCBUSY_MASK) {}
 
-    /* Set sercom module to operate in I2C master mode. */
-    I2CSercom->CTRLA.reg = SERCOM_I2CM_CTRLA_MODE_I2C_MASTER;
+    /* Set sercom module to operate in I2C master mode and run in Standby
+    if user requests it */
+    dev(bus)->CTRLA.reg = SERCOM_I2CM_CTRLA_MODE_I2C_MASTER |
+                          (i2c_config[bus].runstdby ?
+                              SERCOM_I2CM_CTRLA_RUNSTDBY : 0);
 
     /* Enable Smart Mode (ACK is sent when DATA.DATA is read) */
-    I2CSercom->CTRLB.reg = SERCOM_I2CM_CTRLB_SMEN;
+    dev(bus)->CTRLB.reg = SERCOM_I2CM_CTRLB_SMEN;
 
     /* Find and set baudrate. Read speed configuration. Set transfer
      * speed: SERCOM_I2CM_CTRLA_SPEED(0): Standard-mode (Sm) up to 100
      * kHz and Fast-mode (Fm) up to 400 kHz */
     switch (speed) {
```

Comment:
Idea: This could be refactored if the I2C speed enum contains the values which create `tmp_baud`

Reply:
Hum, I understand what you mean. But the calculation requires to know both CORE_CLOCK and the I2C desired speed. I don't think the user must rewrite those values if he changes CORE_CLOCK.
Maybe others maintainers have some ideas ?

Reply:
I meant the following: you can redefine the `i2c_speed_t` in `periph_cpu_common.h` as 

```C
typedef enum {
I2C_SPEED_NORMAL = 100000,
I2C_SPEED_FAST = 400000,
...
} i2c_speed_t;
```

Then you can skip the `switch`:

```C
tmp_baud = (int32_t)(((CLOCK_CORECLOCK + (2 * (speed)) - 1) / (2 * (speed))) - 5);
if (tmp_baud < 255 && tmp_baud > 0) {
	dev(bus)->CTRLA.reg |= SERCOM_I2CM_CTRLA_SPEED(0);
	dev(bus)->BAUD.reg = SERCOM_I2CM_BAUD_BAUD(tmp_baud);
}
```

But now I relised that the `high` speed has different values (2 instead of 0 and HSBAD instead of BAUD), so I'm not sure it makes sense.